# Машинное обучение, ФКН ВШЭ

## Практическое задание 1

### Общая информация
Дата выдачи: 07.09.2018  

Мягкий дедлайн: 07:59MSK 15.09.2018 (за каждый день просрочки снимается 1 балл)

Жесткий дедлайн: 23:59MSK 16.09.2018  

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В каждом разделе вам предлагается выполнить несколько заданий. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотекам pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Задания сдаются через систему Anytask. Инвайт можно найти на странице курса. Присылать необходимо ноутбук с выполненным заданием. 

Для удобства проверки самостоятельно посчитайте свою максимальную оценку (исходя из набора решенных задач) и укажите ниже.

Оценка: xx.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написаем production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [1]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

А это ___ячейка с текстом___.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

Делим целочисленно

In [2]:
5 // 4

1

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. #Делать 
##заголовки 
###разного уровня
3. Выделять *текст* <s>при</s> **необходимости**
4. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных. С этой библиотекой мы сейчас и познакомимся.

Основной объект в pandas это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [3]:
# import almost all we need
%pylab inline 
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Данные можно скачать [отсюда](https://www.dropbox.com/s/5qq94wzmbw4e54r/data.csv?dl=0).

#### 1. [0.5 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [6]:
frame = pd.read_csv('data.csv', header = 0, sep = ',')
print (frame, end="\n\n")
print (f"Имена столбцов: {frame.columns}", end="\n\n")
print (f"Размер таблицы: {frame.shape}", end="\n\n")
print (f"{frame.index}", end="\n\n")

      order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]  

Имеем, таблицу, в которой 4622 строки, исключаяя строку заголовка, а также 5 столбцов. Можно предположить, по названиям столбцов, что это таблица наименований из заказов заведения общепита.

Первый столбец, идентификатор заказа (возможно secondary_key) к которому относится набор данного типа товаров (или один товар).

Второй столбец, количество единиц данного товара в конкретном заказе.

Третий столбец, наименование товара в конкретном заказе.

Четвёртый столбец, описание особенностей состава для конкретного наименования.

Пятый столбец, цена в данном заказе, для заданного количества единиц данного типа продукта, описанного в столбцах третьем и четвёртом.

#### 2. [0.25 баллов] Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [43]:
print(f"Количество уникальных заказов в выборке: {len(set(frame.order_id))}.", end="\n\n")
print(f"Количество уникальных категорий товара: {len(set(frame.item_name))}.")

Количество уникальных заказов в выборке: 1834.

Количество уникальных категорий товара: 50.


#### 3. [0.25 баллов] Есть ли в данных пропуски? В каких колонках? 

In [7]:
print(frame.info(), end="\n\n")

print(f"В данных есть пропуски, в колонке 'choice_description'.")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB
None

В данных есть пропуски, в колонке 'choice_description'.


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [8]:
print("Пустых числовых полей найдено небыло, а строковые были заполнены пустой строкой:", end="\n\n")

frame.choice_description.fillna("", inplace=True)
print(frame.info())

Пустых числовых полей найдено небыло, а строковые были заполнены пустой строкой:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  4622 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB
None


#### 4. [0.5 баллов] Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [10]:
frame["item_num_price"] = frame["item_price"].replace('[\$,]', '', regex=True).astype(float)
print(f"Добавлена новая колонка с вещественными значениями для цен на товары: {frame.item_num_price}")

Добавлена новая колонка с вещественными значениями для цен на товары: 0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_num_price, Length: 4622, dtype: float64


Так как в колонке "item_price" указана цена за набор полностью однотипных товаров по их общей цене, то для удобства анализа сформируем ещё одну колонку "num_price_for_item" в которой отразим цену за один конкретный товар исходя из уникальности по его названию и составу для данного заказа.

In [12]:
frame["num_price_for_item"] = frame.item_num_price / frame.quantity
print(f"Добавлена новая колонка с вещественными значениями для цен на товары: {frame.num_price_for_item}")

Добавлена новая колонка с вещественными значениями для цен на товары: 0        2.39
1        3.39
2        3.39
3        2.39
4        8.49
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: num_price_for_item, Length: 4622, dtype: float64


Какая средняя/минимальная/максимальная цена у товара? 

In [17]:
print(f"Среднее значение цены товара: {round(frame.num_price_for_item.mean(),2)}")
print(f"Минимальное значение цены товара: {frame.num_price_for_item.min()}")
print(f"Максимальное значение цены товара: {frame.num_price_for_item.max()}")

Среднее значение цены товара: 7.08
Минимальное значение цены товара: 1.09
Максимальное значение цены товара: 11.89


Удалите старую колонку с ценой.

In [16]:
frame.drop(['item_price'], axis = 1, inplace = True)
frame

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_num_price  \
0                                                                  2.39   
1                                          [Clementine]            3.39   
2                                               [Apple]            3.39   
3                                                                  2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...           16.98   
...                                                 ...             ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...           11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...           11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...           11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...            8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...            8.75   

      num_price_for_item  
0                   2.39  
1                   3.39  
2                   3.39  
3                   2.39  
4                   8.49  
...                  ...  
4617               11.75  
4618               11.75  
4619               11.25  
4620                8.75  
4621                8.75  

[4622 rows x 6 columns]

#### 5. [0.25 баллов] Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по choice_description)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

Судя по всему, исключать дубликаты следует по уникальности 2х полей: item_name & choice_description. (В данном предположении я исхожу из того, что цены на индивидуальные товары фиксированы в выборке). Из этого будет следовать, что атомарная цена с уникальным item_name & choice_description будет всегда иметь одинаковое значение num_price_for_item, а следовательно, достаточно оставить только первое вхождение такой строки в выборку, а остальные, исключить.

В проецессе выполнения задания, я не буду вносить изменения в выборку, а буду использовать команду, чтобы фильтровать данные "на лету" каждый раз из той выборки, что была изначально.

In [31]:
display("Пять самых дешёвых товаров: ", frame[(~frame.duplicated(subset=["item_name","choice_description"], keep='first'))].sort_values("num_price_for_item").head(15))
display("Пять самых дорогих товаров: ", frame[(~frame.duplicated(subset=["item_name","choice_description"], keep='first'))].sort_values("num_price_for_item").tail(5))

'Пять самых дешёвых товаров: '

order_id  quantity                           item_name  \
51         23         2                         Canned Soda   
34         17         1                       Bottled Water   
117        51         1                         Canned Soda   
126        55         1                         Canned Soda   
28         14         1                         Canned Soda   
162        73         2                         Canned Soda   
18          9         2                         Canned Soda   
263       114         1                   Canned Soft Drink   
368       159         1                   Canned Soft Drink   
610       251         1                   Canned Soft Drink   
346       149         1                   Canned Soft Drink   
292       127         1                   Canned Soft Drink   
6           3         1                       Side of Chips   
40         19         1                               Chips   
191        85         1  Chips and Roasted Chili-Corn Salsa   

    choice_description  item_num_price  num_price_for_item  
51      [Mountain Dew]            2.18                1.09  
34                                1.09                1.09  
117  [Diet Dr. Pepper]            1.09                1.09  
126        [Coca Cola]            1.09                1.09  
28        [Dr. Pepper]            1.09                1.09  
162        [Diet Coke]            2.18                1.09  
18            [Sprite]            2.18                1.09  
263             [Coke]            1.25                1.25  
368        [Diet Coke]            1.25                1.25  
610           [Nestea]            1.25                1.25  
346         [Lemonade]            1.25                1.25  
292           [Sprite]            1.25                1.25  
6                                 1.69                1.69  
40                                2.15                2.15  
191                               2.39                2.39

'Пять самых дорогих товаров: '

order_id  quantity            item_name  \
1505       612         1     Steak Salad Bowl   
3350      1343         1     Steak Salad Bowl   
4239      1692         1  Carnitas Salad Bowl   
2442       970         1  Barbacoa Salad Bowl   
1816       734         1     Steak Salad Bowl   

                                     choice_description  item_num_price  \
1505  [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...           11.89   
3350  [Fresh Tomato Salsa, [Cheese, Guacamole, Lettu...           11.89   
4239  [Tomatillo Green Chili Salsa, [Black Beans, Ch...           11.89   
2442  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...           11.89   
1816  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...           11.89   

      num_price_for_item  
1505               11.89  
3350               11.89  
4239               11.89  
2442               11.89  
1816               11.89

In [34]:
display("Количество уникальных товаров: ", len(frame[(~frame.duplicated(subset=["item_name","choice_description"], keep='first'))]))

'Количество уникальных товаров: '

1871

#### 6. [0.5 баллов] Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [ ]:
# your code

#### 7. [0.5 баллов] Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [ ]:
# your code

#### 8. [0.25 баллов] Сколько заказов содержали ровно 1 товар?

In [ ]:
# your code

#### 9. [0.25 баллов] Какая самая популярная категория товара? 

In [ ]:
# your code

#### 10. [0.5 баллов] Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой? 

In [ ]:
# your code

#### 11. [0.75 баллов] В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно средний чек, добавить его в исходные данные и сделать необходимые проверки.

In [ ]:
# your code

#### 12. [0.75 баллов] Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [ ]:
# your code

#### 13. [0.75 баллов] Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [ ]:
data = np.random.rand(10, 3)

# your code

Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [ ]:
# your code

## 2. Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).

In [20]:
# нужно для отображения графиков внутри ноутбука
%matplotlib inline
import matplotlib.pyplot as plt

На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [ ]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])

#### 1. [1 балл] Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name. 

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [ ]:
# your code

#### 2. [1 балл] Постройте график зависимости суммы покупок от дней.

In [ ]:
# your code

#### 3. [1 балл] Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
# your code

#### 4. [1 балл] Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Кстати, существует надстройка над matplotlib под названием [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Иногда удобнее и красивее делать визуализации через неё. 